segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
-     * @param chunksToDelete List of chunks to delete.
-     */
-    void addToGarbage(Collection<String> chunksToDelete) {
-        val currentTime = currentTimeSupplier.get();
-
-        chunksToDelete.forEach(chunkToDelete -> addToGarbage(chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0));
+    CompletableFuture<Void>  trackNewChunk(long transactionId, String chunktoTrack) {
+        if (null != taskQueue) {
+            val startTime = currentTimeSupplier.get() + config.getGarbageCollectionDelay().toMillis();
+            return taskQueue.addTask(taskQueueName, new TaskInfo(chunktoTrack, startTime, 0, TaskInfo.DELETE_CHUNK, transactionId))
+                    .thenRunAsync(() -> {
+                        queueSize.incrementAndGet();
+                        SLTS_GC_CHUNK_NEW.inc();
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
+    }
 
-        if (queueSize.get() >= config.getGarbageCollectionMaxQueueSize()) {
-            log.warn("{}: deleteGarbage - Queue full. Could not delete garbage. Chunks skipped", traceObjectId);
+    private CompletableFuture<Void> failTask(TaskInfo infoToRetire) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(failedQueueName, infoToRetire);
         }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private CompletableFuture<Void> deleteSegment(TaskInfo taskInfo) {
+        val streamSegmentName = taskInfo.getName();
+        ArrayList<String> chunksToDelete = new ArrayList<>();
+        @Cleanup
+        val txn = metadataStore.beginTransaction(false, streamSegmentName);
+        return txn.get(streamSegmentName)
+                .thenComposeAsync(storageMetadata -> {
+                    val segmentMetadata = (SegmentMetadata) storageMetadata;
+                    val failed = new AtomicReference<Throwable>();
+                    if (null == segmentMetadata) {
+                        log.debug("{}: deleteGarbage - Segment metadata does not exist. segment={}.", traceObjectId, streamSegmentName);
+                        return CompletableFuture.completedFuture(null);
+                    } else if (segmentMetadata.isActive()) {
+                        log.debug("{}: deleteGarbage - Segment is not marked as deleted. segment={}.", traceObjectId, streamSegmentName);
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return new ChunkIterator(storageExecutor, txn, segmentMetadata)
+                                .forEach((metadata, name) -> {
+                                    metadata.setActive(false);
+                                    txn.update(metadata);
+                                    chunksToDelete.add(name);
+                                })
+                                .thenRunAsync(() -> this.addChunksToGarbage(txn.getVersion(), chunksToDelete), storageExecutor)
```

Comment:
Combine these two thenRun + the next thenCompose into a single method.

Reply:
These are two separate steps.
Also this is background processing task. Releasing thread pool thread intermittently like this would be actually benefetial